### PR TITLE
Introducing es alias fields

### PR DESCRIPTION
**Pull request by iagodvsantos**
# What

We've introduced a new feature for field aliasing this way search-api will now be able to (filter/query/sort) using an alias for a field which is loaded from `search-api-properties`

# How was this patch tested?

Unit and integration tests and faith :pray:

ref:. (https://github.com/grupozap/squad-search-ranking/issues/1349)

### CODE REVIEW

File: src/main/java/com/grupozap/search/api/model/parser/FieldParser.java [hunk 1]
```diff
@@ -18,22 +19,24 @@
   private final Parser<Field> fieldParserWithNot;
 
   public FieldParser(NotParser notParser) {
-    this(notParser, FieldFactory::createField, FieldFactory::createField);
+    this(notParser, fieldName -> fieldName, FieldFactory::createField, FieldFactory::createField);
```
Comment (rtokoi) [inline src/main/java/com/grupozap/search/api/model/parser/FieldParser.java]: :q-delicia-cara:

File: src/integration-test/java/com/grupozap/search/api/itest/scenarios/SearchIntegrationTest.java [hunk 2]
```diff
@@ -1369,4 +1369,54 @@ public void validateSearchUsingPolygonWithLessThan3Points() {
...
+    var id = standardDatasetSize / 2;
+    given()
+        .log()
+        .all()
+        .baseUri(baseUrl)
+        .contentType(JSON)
+        .expect()
+        .statusCode(SC_BAD_REQUEST)
+        .when()
+        .get(format("%s/%s?includeFields=field_before_alias", TEST_DATA_INDEX, id));
```
Comment (mezzatto) [inline src/integration-test/java/com/grupozap/search/api/itest/scenarios/SearchIntegrationTest.java]: `excludeFields` has the same behavior, right?

File: src/main/java/com/grupozap/search/api/service/parser/factory/ESAlias.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+  public Map<String, ESAliasProp> getAliases() {
+    return this.aliases;
+  }
+
+  @Override
+  public void onApplicationEvent(RemotePropertiesUpdatedEvent event) {
+    Object value = ES_ALIAS.getValue(event.getIndex());
+    if (nonNull(value)) {
+      this.aliases.put(event.getIndex(), convertValue(value, ESAliasProp.class));
```
Comment (mezzatto) [inline src/main/java/com/grupozap/search/api/service/parser/factory/ESAlias.java]: What happens when you create an alias to a field that does not exist? For instance, `my_alias` -> `mezzatto_field`

File: src/integration-test/java/com/grupozap/search/api/itest/scenarios/SearchIntegrationTest.java (hunk 2 above)
Comment (iagodvsantos) [reply src/integration-test/java/com/grupozap/search/api/itest/scenarios/SearchIntegrationTest.java]: Yeah, it's the same behavior, I will add a test for that scenario

File: src/main/java/com/grupozap/search/api/service/parser/factory/ESAlias.java (hunk 3 above)
Comment (iagodvsantos) [reply src/main/java/com/grupozap/search/api/service/parser/factory/ESAlias.java]: Currently, the request will fail saying the field does not exist, it's possible to add a validation to ensure that alias must be mapped to existing fields this way application wouldn't even start. What do you think?

Comment (mezzatto) [reply src/main/java/com/grupozap/search/api/service/parser/factory/ESAlias.java]: This looks dangerous. Better to keep the way it is 👍

File: src/main/java/com/grupozap/search/api/service/parser/factory/ESAlias.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+  private static final Logger LOG = LoggerFactory.getLogger(ESAlias.class);
+
+  private final IndexSettings indexSettings;
+  private Map<String, ESAliasProp> aliases;
+
+  @Autowired
+  public ESAlias(IndexSettings indexSettings) {
+    this.indexSettings = indexSettings;
+    this.aliases = new HashMap<>();
```
Comment (marciorodrigues87) [inline src/main/java/com/grupozap/search/api/service/parser/factory/ESAlias.java]: `HashMap` is not Threadsafe, can you change it to a `ConcurrentHashMap` please?

File: src/main/java/com/grupozap/search/api/service/parser/factory/ESAlias.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+  public String getIndexAlias(String indexName) {
+    return this.aliases
+        .get(indexSettings.getIndex())
+        .getIndices()
+        .getOrDefault(indexName, indexName);
+  }
+
+  public static final class ESAliasProp {
+    private Map<String, String> fields;
+    private Map<String, String> indices;
```
Comment (marciorodrigues87) [inline src/main/java/com/grupozap/search/api/service/parser/factory/ESAlias.java]: Typo, `indexes`

File: src/main/java/com/grupozap/search/api/service/parser/factory/ESAlias.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+    ESAliasProp() {
+      this.fields = new HashMap<>();
```
Comment (marciorodrigues87) [inline src/main/java/com/grupozap/search/api/service/parser/factory/ESAlias.java]: Better use `ConcurrentHashMap` here too

File: src/main/java/com/grupozap/search/api/service/parser/factory/ESAlias.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+      this.indices = new HashMap<>();
```
Comment (marciorodrigues87) [inline src/main/java/com/grupozap/search/api/service/parser/factory/ESAlias.java]: Here too

File: src/main/java/com/grupozap/search/api/service/parser/factory/ESAlias.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+import java.util.HashMap;
+import java.util.Map;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.ApplicationListener;
+import org.springframework.stereotype.Component;
+
+@Component("esAlias")
+public class ESAlias implements ApplicationListener<RemotePropertiesUpdatedEvent> {
```
Comment (marciorodrigues87) [inline src/main/java/com/grupozap/search/api/service/parser/factory/ESAlias.java]: I know it's just a name, but `ES` is directly related to `Elastic Search` and `search-api` is meant to be agnostic of search engine, so I think we can think in another name for this class 💅

File: src/main/java/com/grupozap/search/api/service/parser/factory/ESAlias.java (hunk 5 above)
Comment (iagodvsantos) [reply src/main/java/com/grupozap/search/api/service/parser/factory/ESAlias.java]: Actually, both `indexes` and `indices` are correct forms

https://business.nasdaq.com/marketinsite/2016/indexes-or-indices-whats-the-deal.html

File: src/main/java/com/grupozap/search/api/service/parser/factory/ESAlias.java (hunk 8 above)
Comment (iagodvsantos) [reply src/main/java/com/grupozap/search/api/service/parser/factory/ESAlias.java]: Good point :heart:

**Comment by iagodvsantos** [comment]
Could you review again @marciorodrigues87, please?

**Pull request merged**